File: client/src/modules/nexotype/schemas/user/user-treatment-log.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const UserTreatmentLogSchema = z.object({
  id: z.number(),
  subject_id: z.number().int(),
  asset_id: z.number().int(),
  dosage: z.string().min(1).max(100),
  started_at: z.string(),
  ended_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateUserTreatmentLogSchema = z.object({
  subject_id: z.number().int(),
  asset_id: z.number().int(),
  dosage: z.string().min(1).max(100),
  started_at: z.string().min(1),
  ended_at: z.string().nullable().optional(),
});

export const UpdateUserTreatmentLogSchema = z.object({
  subject_id: z.number().int().optional(),
  asset_id: z.number().int().optional(),
  dosage: z.string().min(1).max(100).optional(),
  started_at: z.string().optional(),
  ended_at: z.string().nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type UserTreatmentLog = z.infer<typeof UserTreatmentLogSchema>;
export type CreateUserTreatmentLog = z.infer<typeof CreateUserTreatmentLogSchema>;
export type UpdateUserTreatmentLog = z.infer<typeof UpdateUserTreatmentLogSchema>;

// ==========================================
// Response Types
// ==========================================

export type UserTreatmentLogResponse = {
  success: boolean;
  data?: UserTreatmentLog;
  error?: string;
};

export type UserTreatmentLogsResponse = {
  success: boolean;
  data?: UserTreatmentLog[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/user/user_treatment_log_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import date, datetime

# ==========================================
# Request Schemas
# ==========================================

class UserTreatmentLogCreate(BaseModel):
    """Schema for creating a new user treatment log"""
    subject_id: int
    asset_id: int
    dosage: str
    started_at: date
    ended_at: date | None = None


class UserTreatmentLogUpdate(BaseModel):
    """Schema for updating a user treatment log"""
    subject_id: int = None
    asset_id: int = None
    dosage: str = None
    started_at: date = None
    ended_at: date | None = None


# ==========================================
# Response Schemas
# ==========================================

class UserTreatmentLogDetail(BaseModel):
    """Schema for user treatment log details"""
    id: int
    subject_id: int
    asset_id: int
    dosage: str
    started_at: date
    ended_at: date | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class UserTreatmentLogListResponse(BaseModel):
    """Response schema for listing user treatment logs"""
    success: bool
    data: list[UserTreatmentLogDetail] | None = None
    count: int = 0
    error: str | None = None


class UserTreatmentLogResponse(BaseModel):
    """Response schema for single user treatment log operations"""
    success: bool
    data: UserTreatmentLogDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
